crates/bazed-core/src/vim_interface.rs

```diff
@@ -0,0 +1,114 @@
+use bazed_rpc::keycode::{Key, KeyInput};
+
+use crate::{
+    buffer::Buffer,
+    user_buffer_op::{BufferOp, Motion, Trajectory},
+    view::View,
+    word_boundary::WordBoundaryType,
+};
+
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, derive_more::Display)]
+enum VimMode {
+    #[default]
+    Normal,
+    Insert,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct VimInterface {
+    mode: VimMode,
+}
+
+impl VimInterface {
+    pub(crate) fn on_input(&mut self, view: &View, buffer: &mut Buffer, input: KeyInput) {
+        match self.mode {
```

Comment:
This should probably be some sort of &dyn InputThing in general, so we can swap them out, and get Rust to dispatch.

Reply:
Might be a good refactor in the future, however I think for now this is sufficient, and we'll need to consider things like macros, which have state that stays relevant across mode changes